packages/upgrades/src/upgradeable.cairo

```diff
@@ -42,8 +43,30 @@ pub mod UpgradeableComponent {
         /// Emits an `Upgraded` event.
         fn upgrade(ref self: ComponentState<TContractState>, new_class_hash: ClassHash) {
             assert(!new_class_hash.is_zero(), Errors::INVALID_CLASS);
-            starknet::syscalls::replace_class_syscall(new_class_hash).unwrap_syscall();
+            replace_class_syscall(new_class_hash).unwrap_syscall();
             self.emit(Upgraded { class_hash: new_class_hash });
         }
+
+        /// Replaces the contract's class hash with `new_class_hash` and then calls `selector`
+        /// from the upgraded context.
+        ///
+        /// Requirements:
+        ///
+        /// - `new_class_hash` is not zero.
+        ///
+        /// Emits an `Upgraded` event.
+        fn upgrade_and_call(
+            ref self: ComponentState<TContractState>,
+            new_class_hash: ClassHash,
+            selector: felt252,
+            calldata: Span<felt252>
+        ) {
+            self.upgrade(new_class_hash);
+            let this = starknet::get_contract_address();
+            // `call_contract_syscall` is used in order to call `selector` from the new class.
+            // See:
+            // https://docs.starknet.io/documentation/architecture_and_concepts/Contracts/system-calls-cairo1/#replace_class
+            call_contract_syscall(this, selector, calldata).unwrap_syscall();
+        }
```

Comment:
Agreed! Will update

Reply:
Not unwrapping the syscall has an important drawback: the user may think the call was successful when it wasn't if he fails to unwrap the call himself. As one of our priorities is to make the library as hard to misuse as possible, I vote we keep the unwrap syscall. If the user has a more complex flow requiring the Result, they can and should implement the mechanism, to ensure they are aware of the implication.

Reply:
Good point. Not unwrapping can be an easy oversight. I was wondering if it'd make sense to have an internal "unsafe" `_upgrade_and_call` that returns the res. It's probably better to not take on the responsibility if we don't have to and instead have users implement it themselves as Eric said. The normal `upgrade` and `upgrade_and_call` are already internal so this hinges on the prefixed underscore. It is an option though if we want the flexibility

Reply:
I'm ok with adding an "unsafe" version.

Reply:
Wait, why can't we unwrap the syscall and just return the result?

```
fn upgrade_and_call(
    ref self: ComponentState<TContractState>,
    new_class_hash: ClassHash,
    selector: felt252,
    calldata: Span<felt252>
) -> Span<felt252> {
    self.upgrade(new_class_hash);
    let this = starknet::get_contract_address();
    call_contract_syscall(this, selector, calldata).unwrap_syscall()
}
```

Reply:
I toughy @immrsd was referring to returning the Result object, in case the user wants to use that. But I noticed now that it may be referring to the return value when the result is Ok, and then I'm ok with just returning that as you are pointing up. I like that better than returning the Result object.

Reply:
Note that we need to update the IUpgradeAndCall interface ID from this change.

Reply:
Yes, that was the idea